From 17720d7b36199025b01a1dc0b5a8e63b13857783 Mon Sep 17 00:00:00 2001
From: Girija P <gpandiya@qti.qualcomm.com>
Date: Mon, 25 May 2026 11:30:10 +0530
Subject: [PATCH] Update MLD BSTA related configurations from M2 WSC and BSTA
 MLD configuration message to wireless uci config for NLHAL enabled platforms

Signed-off-by: Girija P <gpandiya@qti.qualcomm.com>
---
 agent/src/beerocks/slave/agent_db.h           |  10 +
 .../backhaul_manager/backhaul_manager.cpp     |  71 ++-
 .../slave/backhaul_manager/backhaul_manager.h |  14 +
 .../slave/tasks/ap_autoconfiguration_task.cpp |  25 +-
 .../beerocks/bcl/include/bcl/beerocks_utils.h |  14 +
 common/beerocks/bcl/source/beerocks_utils.cpp |  20 +
 .../bwl/nl80211/sta_wlan_hal_nl80211.cpp      |  13 +-
 .../platform/bpl/cfg/dm/bpl_cfg_wifi.cpp      |  11 +-
 framework/platform/bpl/cfg/rdk/bpl_cfg.cpp    |  12 +-
 .../platform/bpl/cfg/uci/bpl_cfg_wifi.cpp     | 583 +++++++++++++++---
 framework/platform/bpl/include/bpl/bpl_cfg.h  |  29 +-
 11 files changed, 694 insertions(+), 108 deletions(-)

diff --git a/agent/src/beerocks/slave/agent_db.h b/agent/src/beerocks/slave/agent_db.h
index f38777597..81ddbdcf7 100644
--- a/agent/src/beerocks/slave/agent_db.h
+++ b/agent/src/beerocks/slave/agent_db.h
@@ -772,6 +772,16 @@ public:
     std::vector<sAPMLDConfiguration> ap_mld_configurations;
     std::unique_ptr<sBStaMLDConfiguration> bsta_mld_configuration;
 
+    /**
+     * @brief Flag indicating that wpa_supplicant restart is required.
+     *
+     * Set to true when bpl_cfg_update_bsta_mld_grp() returns
+     * eBstaSupplicantRestartStatus::SUPP_RST_REQUIRED, indicating that the
+     * MLD group option was updated in UCI and the supplicant must be restarted
+     * to pick up the new configuration.
+     */
+    bool supplicant_restart_required = false;
+
     /**
      * @brief Map of associated MLO (Multi-Link Operation) client stations.
      *
diff --git a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
index 12eb55471..f11a782b9 100644
--- a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
+++ b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
@@ -2383,11 +2383,22 @@ bool BackhaulManager::handle_slave_backhaul_message(int fd, ieee1905_1::CmduMess
         if (!db->device_conf.local_gw) {
 
             for (auto &radio_info : m_radios_info) {
+/*
                 std::string iface = radio_info->sta_iface;
                 if (radio_info->sta_iface.empty()) {
                     LOG(DEBUG) << " STA Iface is empty for radio: " << radio_mac << " skip ACTION_BACKHAUL_WIFI_CREDENTIALS_UPDATE_REQUEST " ;
                     continue;
                 }
+*/
+
+                // Retrieve the radio band (radio_band0/radio_band1/radio_band2) for this radio MAC
+                auto radio_band = get_radio_band(radio_mac);
+                if (radio_band.empty()) {
+                    LOG(ERROR) << "Failed to get radio band for radio " << radio_mac;
+                    return false;
+                }
+                LOG(DEBUG) << "Radio " << radio_mac << " belongs to band: " << radio_band;
+
                 if (radio_info->radio_mac == radio_mac) {
                     LOG(DEBUG) << " Radio matched ";
                     auto config_data = request->wifi_credentials();
@@ -2413,7 +2424,7 @@ bool BackhaulManager::handle_slave_backhaul_message(int fd, ieee1905_1::CmduMess
  */
 #ifdef USE_PRPLMESH_NLHAL
                         // Write credentials to /etc/config/wireless for persistency
-                        if (!beerocks::bpl::bpl_cfg_set_sta_vap_bh_credentials(radio_info->sta_iface, radio_info->bSta_conf)) {
+                        if (!beerocks::bpl::bpl_cfg_set_sta_vap_bh_credentials(radio_info->sta_iface, radio_info->bSta_conf, get_radio_band(radio_info->radio_mac))) {
                             LOG(WARNING) << "Failed to write backhaul credentials to wireless config for " << radio_info->sta_iface;
                         } else {
                             LOG(DEBUG) << "Successfully wrote backhaul credentials to wireless config for " << radio_info->sta_iface;
@@ -2437,12 +2448,52 @@ bool BackhaulManager::handle_slave_backhaul_message(int fd, ieee1905_1::CmduMess
 
         const sMacAddr &radio_mac = request->radio_mac();
 
+/* Still continuing with single sta_wlan_hal attach for MLD BSTA, until MLD BSTA is enabled in community,
+   As sta_wlan_hal is not enabled to be unique for each affiliated sta in current arch, once enabled
+   ifdefs can be removed and introduced changes can be generalised.*/
+
         LOG(DEBUG) << "handle ACTION_BACKHAUL_MLD_UPDATE_REQUEST for bsta on radio " << radio_mac;
+
+        // Retrieve the radio band (radio_band0/radio_band1/radio_band2) for this radio MAC
+        auto radio_band = get_radio_band(radio_mac);
+        if (radio_band.empty()) {
+            LOG(ERROR) << "Failed to get radio band for radio " << radio_mac;
+            return false;
+        }
+        LOG(DEBUG) << "Radio " << radio_mac << " belongs to band: " << radio_band;
+
+#ifdef USE_PRPLMESH_NLHAL
+        auto restart_status =
+            beerocks::bpl::bpl_cfg_update_bsta_mld_grp(radio_band, request->mld_unit());
+        if (restart_status ==
+            beerocks::bpl::eBstaSupplicantRestartStatus::SUPP_RST_REQUIRED) {
+            auto db                        = AgentDB::get();
+            db->supplicant_restart_required = true;
+            LOG(DEBUG) << "Supplicant restart required after MLD group update for radio "
+                       << radio_mac;
+        }
+#endif
+
         auto sta_wlan_hal = get_wireless_hal(radio_mac);
         if (!sta_wlan_hal) {
             LOG(ERROR) << "Failed to get wireless hal for radio " << radio_mac;
             return false;
         }
+
+/* if sta_wlan_hal is in attached state or if connect is tried somewhere later, check if supp_rst flag is present
+ * in sta_wlan_hal, if so, restart the supplicant, wait for a timeout and 
+ * then try attach or connect anywhere from backhaul manager 
+ * This makes sure config is saved during ethernet backhaul too and when eth to wifi switching happens,
+ * new mld config will be effective and can be used */
+#ifdef USE_PRPLMESH_NLHAL
+        auto db                        = AgentDB::get();
+        if (db->supplicant_restart_required) {
+            LOG(DEBUG) << " Supplicant restart will be triggered by STA WLAN HAL this time\n";
+            db->supplicant_restart_required = false;
+        } else {
+            LOG(DEBUG) << " Supplicant restart will not be triggered by STA WLAN HAL this time\n";
+        }
+#endif
         sta_wlan_hal->update_mld_mode(request->mld_mode());
         sta_wlan_hal->update_mld_unit(request->mld_unit());
         break;
@@ -3372,6 +3423,24 @@ std::shared_ptr<bwl::sta_wlan_hal> BackhaulManager::get_wireless_hal(const sMacA
     return (*radio_it)->sta_wlan_hal;
 }
 
+std::string BackhaulManager::get_radio_band(const sMacAddr &radio_mac)
+{
+    auto db    = AgentDB::get();
+    auto radio = db->get_radio_by_mac(radio_mac, AgentDB::eMacType::RADIO);
+    if (!radio) {
+        LOG(ERROR) << "get_radio_band: radio not found in AgentDB for MAC " << radio_mac;
+        return {};
+    }
+
+    const auto freq_type  = radio->wifi_channel.get_freq_type();
+    const auto radio_band = beerocks::utils::freq_type_to_radio_band_name(freq_type);
+    if (radio_band.empty()) {
+        LOG(ERROR) << "get_radio_band: unknown frequency type " << int(freq_type)
+                   << " for radio " << radio_mac;
+    }
+    return radio_band;
+}
+
 bool BackhaulManager::handle_slave_failed_connection_message(ieee1905_1::CmduMessageRx &cmdu_rx,
                                                              const sMacAddr &src_mac)
 {
diff --git a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.h b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.h
index 7491e855b..4e97e66f7 100644
--- a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.h
+++ b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.h
@@ -310,6 +310,20 @@ private:
     std::shared_ptr<bwl::sta_wlan_hal> get_wireless_hal(std::string iface = "");
     std::shared_ptr<bwl::sta_wlan_hal> get_wireless_hal(const sMacAddr &radio_mac);
 
+    /**
+     * @brief Get the radio band name (radio_band0/radio_band1/radio_band2) for a given radio MAC.
+     *
+     * Looks up the radio in AgentDB by its MAC address and maps its frequency type to a
+     * band name string:
+     *   - FREQ_24G              -> "radio_band0"
+     *   - FREQ_5G/5GL/5GH      -> "radio_band1"
+     *   - FREQ_6G               -> "radio_band2"
+     *
+     * @param radio_mac MAC address of the radio.
+     * @return std::string Band name string, or empty string on failure.
+     */
+    std::string get_radio_band(const sMacAddr &radio_mac);
+
     /**
      * @brief Converts WSC::sWscAttrAuthenticationType to bwl::WiFiSec
      */
diff --git a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
index 2bf457342..67e96e9bc 100644
--- a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
+++ b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
@@ -2767,14 +2767,31 @@ bool ApAutoConfigurationTask::handle_bsta_mld_configuration_tlv(ieee1905_1::Cmdu
     // Lambda to process missing MLD links for a given ruid
     auto process_missing_mld_links = [&](const sMacAddr &bsta_ruid) {
         // Find the specific radio in curr_bsta_mld_infos
-        auto it = curr_bsta_mld_infos.find(bsta_ruid);
+        auto it          = curr_bsta_mld_infos.find(bsta_ruid);
+        bool in_old_conf = (it != curr_bsta_mld_infos.end());
+        bool in_new_conf = (bsta_mld_requests_infos.find(bsta_ruid) != bsta_mld_requests_infos.end());
 
-        // If found in old config but not in new config, disable it
-        if (it != curr_bsta_mld_infos.end() &&
-            bsta_mld_requests_infos.find(bsta_ruid) == bsta_mld_requests_infos.end()) {
+        LOG(DEBUG) << "process_missing_mld_links: ruid=" << tlvf::mac_to_string(bsta_ruid)
+                   << " in_old_conf=" << in_old_conf << " in_new_conf=" << in_new_conf;
+
+        if (in_old_conf) {
+            LOG(DEBUG) << "  old_conf: mld_unit=" << int(it->second.first)
+                       << " mld_mode=0x" << std::hex << int(it->second.second) << std::dec;
+        } else {
+            LOG(DEBUG) << "  ruid not found in old bsta_mld_requests_infos"
+                       << " => mld_id -1 will NOT be sent (radio was never configured for BSTA MLD)";
+        }
+
+        // Send DISABLED_MLDUNIT for any radio not present in the new config,
+        // regardless of whether it was in the old config or not.
+        if (!in_new_conf) {
+            LOG(DEBUG) << "  ruid not in new config => sending DISABLED_MLDUNIT"
+                       << (in_old_conf ? " (was in old config)" : " (was not in old config either)");
             send_bsta_mld_configuration(
                 bsta_ruid, DISABLED_MLDUNIT,
                 static_cast<uint8_t>(AgentDB::sMLDConfiguration::mode::NONE));
+        } else {
+            LOG(DEBUG) << "  ruid is in new config => no disable needed";
         }
     };
 
diff --git a/common/beerocks/bcl/include/bcl/beerocks_utils.h b/common/beerocks/bcl/include/bcl/beerocks_utils.h
index e6383c0a6..d9c6ea20d 100644
--- a/common/beerocks/bcl/include/bcl/beerocks_utils.h
+++ b/common/beerocks/bcl/include/bcl/beerocks_utils.h
@@ -106,6 +106,20 @@ public:
     static std::string convert_frequency_type_to_string(beerocks::eFreqType freq_type);
     static std::string convert_channel_ext_above_to_string(bool channel_ext_above_secondary,
                                                            beerocks::eWiFiBandwidth bandwidth);
+
+    /**
+     * @brief Converts a frequency type to a prplmesh radio band name string.
+     *
+     * Maps eFreqType values to the radio band name used in the prplmesh UCI
+     * configuration (e.g. the "radio_band" option in the wifi-device section):
+     *   - FREQ_24G              -> "radio_band0"
+     *   - FREQ_5G / 5GL / 5GH  -> "radio_band1"
+     *   - FREQ_6G               -> "radio_band2"
+     *
+     * @param[in] freq_type Frequency type to convert.
+     * @return Radio band name string, or empty string for unknown/unsupported types.
+     */
+    static std::string freq_type_to_radio_band_name(beerocks::eFreqType freq_type);
     static void hex_dump(const std::string &description, uint8_t *addr, int len,
                          const char *calling_file = __builtin_FILE(),
                          int calling_line         = __builtin_LINE());
diff --git a/common/beerocks/bcl/source/beerocks_utils.cpp b/common/beerocks/bcl/source/beerocks_utils.cpp
index 0f727da12..5d2c6ce42 100644
--- a/common/beerocks/bcl/source/beerocks_utils.cpp
+++ b/common/beerocks/bcl/source/beerocks_utils.cpp
@@ -287,6 +287,26 @@ std::string utils::convert_frequency_type_to_string(beerocks::eFreqType freq_typ
     }
 }
 
+std::string utils::freq_type_to_radio_band_name(beerocks::eFreqType freq_type)
+{
+    switch (freq_type) {
+    case beerocks::FREQ_24G:
+        return "2g";
+    case beerocks::FREQ_5G:
+        return "5g";
+    case beerocks::FREQ_5GL:
+        return "5gl";
+    case beerocks::FREQ_5GH:
+        return "5gh";
+    case beerocks::FREQ_6G:
+        return "6g";
+    default:
+        LOG(WARNING) << "freq_type_to_radio_band_name: unsupported freq_type "
+                     << int(freq_type) << ", returning empty string";
+        return {};
+    }
+}
+
 std::string utils::convert_channel_ext_above_to_string(bool channel_ext_above_secondary,
                                                        beerocks::eWiFiBandwidth bandwidth)
 {
diff --git a/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.cpp
index 0477d4ee3..7cd4c2a31 100644
--- a/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.cpp
@@ -10,6 +10,7 @@
 
 #include <bcl/beerocks_utils.h>
 #include <bcl/network/network_utils.h>
+#include <bpl/bpl_cfg.h>
 
 #include <easylogging++.h>
 
@@ -766,7 +767,17 @@ bool sta_wlan_hal_nl80211::add_new_network(son::wireless_utils::sBssInfoConf bSt
     LOG(DEBUG) << "Network with id " << network_id << " has been added and enabled on interface "
                << sta_iface;
 
-    beerocks::bpl::bpl_cfg_set_sta_vap_bh_credentials(sta_iface,bSta_conf);
+    // Derive radio_band string using get_radio_info().frequency_band and the
+    // beerocks::utils::freq_type_to_radio_band_name() utility function
+    const std::string radio_band =
+        beerocks::utils::freq_type_to_radio_band_name(get_radio_info().frequency_band);
+    if (radio_band.empty()) {
+        LOG(WARNING) << "add_new_network: could not determine radio_band for interface "
+                     << sta_iface << " (frequency_band="
+                     << int(get_radio_info().frequency_band) << ")";
+    }
+
+    beerocks::bpl::bpl_cfg_set_sta_vap_bh_credentials(sta_iface, bSta_conf, radio_band);
 
     return true;
 }
diff --git a/framework/platform/bpl/cfg/dm/bpl_cfg_wifi.cpp b/framework/platform/bpl/cfg/dm/bpl_cfg_wifi.cpp
index d429c8751..b0133f299 100644
--- a/framework/platform/bpl/cfg/dm/bpl_cfg_wifi.cpp
+++ b/framework/platform/bpl/cfg/dm/bpl_cfg_wifi.cpp
@@ -466,7 +466,8 @@ bool cfg_commit_wireless_config()
  */
 /*
 bool bpl_cfg_set_sta_vap_bh_credentials(const std::string &iface,
-                                        const son::wireless_utils::sBssInfoConf &configuration)
+                                        const son::wireless_utils::sBssInfoConf &configuration,
+                                        const std::string &radio_band)
 {
     // For DM/WHM implementation, update the WiFi data model with backhaul credentials
     std::string wifi_ssid_path = wbapi_utils::search_path_ssid_by_iface(iface);
@@ -538,5 +539,13 @@ bool cfg_set_wireless_channel(const std::string &radio_id, int channel)
     return true;
 }
 
+eBstaSupplicantRestartStatus bpl_cfg_update_bsta_mld_grp(const std::string &radio_band,
+                                                         int8_t mld_unit)
+{
+    (void)radio_band;
+    (void)mld_unit;
+    return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
+}
+
 } // namespace bpl
 } // namespace beerocks
diff --git a/framework/platform/bpl/cfg/rdk/bpl_cfg.cpp b/framework/platform/bpl/cfg/rdk/bpl_cfg.cpp
index 69bbe4e7c..41e94fc3d 100644
--- a/framework/platform/bpl/cfg/rdk/bpl_cfg.cpp
+++ b/framework/platform/bpl/cfg/rdk/bpl_cfg.cpp
@@ -824,13 +824,23 @@ bool bpl_cfg_set_wifi_credentials(const std::string &iface,
 }
 
 bool bpl_cfg_set_sta_vap_bh_credentials(const std::string &iface,
-                                        const son::wireless_utils::sBssInfoConf &configuration)
+                                        const son::wireless_utils::sBssInfoConf &configuration,
+                                        const std::string &radio_band)
 {
     (void)iface;
     (void)configuration;
+    (void)radio_band;
     return true;
 }
 
+eBstaSupplicantRestartStatus bpl_cfg_update_bsta_mld_grp(const std::string &radio_band,
+                                                         int8_t mld_unit)
+{
+    (void)radio_band;
+    (void)mld_unit;
+    return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
+}
+
 bool bpl_cfg_get_mandatory_interfaces(std::string &mandatory_interfaces)
 {
     mandatory_interfaces.clear();
diff --git a/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp b/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp
index 44fa58eda..53123d692 100644
--- a/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp
+++ b/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp
@@ -1162,150 +1162,535 @@ bool cfg_set_wifi_iface_network(const std::string &iface_name, const std::string
 
 bool cfg_commit_wireless_config() { return uci_commit_changes("wireless"); }
 
+// ---------------------------------------------------------------------------
+// File-scope helpers shared by bpl_cfg_set_sta_vap_bh_credentials and
+// bpl_cfg_update_bsta_mld_grp
+// ---------------------------------------------------------------------------
+
+/**
+ * @brief Return the refined radio-band string for a wifi-device section.
+ *
+ * Reads the "band" option and refines "5g" to "5gl" or "5gh" based on the
+ * "channels" option.  Returns an empty string when the band cannot be read.
+ */
+static std::string get_refined_band(const std::string &device_section)
+{
+    std::string band;
+    if (!uci_get_option("wireless", "wifi-device", device_section, "band", band)) {
+        return "";
+    }
+    if (band == "5g") {
+        std::string channel_range;
+        if (uci_get_option("wireless", "wifi-device", device_section, "channels",
+                           channel_range)) {
+            if (channel_range == "36-64") {
+                band = "5gl";
+            } else if (channel_range == "100-165") {
+                band = "5gh";
+            }
+        }
+    }
+    return band;
+}
+
+/**
+ * @brief Convert WSC authentication/encryption types to a UCI encryption string.
+ */
+static std::string get_encryption_string(WSC::eWscAuth authentication_type,
+                                         WSC::eWscEncr encryption_type)
+{
+    std::string encryption = "none";
+    if (authentication_type == WSC::eWscAuth::WSC_AUTH_WPA2PSK) {
+        encryption = "psk2";
+        if (encryption_type == WSC::eWscEncr::WSC_ENCR_TKIP) {
+            encryption += "+tkip";
+        } else if (encryption_type == WSC::eWscEncr::WSC_ENCR_AES) {
+            encryption += "+aes";
+        }
+    } else if (authentication_type == WSC::eWscAuth::WSC_AUTH_SAE) {
+        encryption = "sae";
+    } else if (authentication_type ==
+               static_cast<WSC::eWscAuth>(WSC::eWscAuth::WSC_AUTH_SAE |
+                                          WSC::eWscAuth::WSC_AUTH_WPA2PSK)) {
+        encryption = "sae-mixed";
+    }
+    return encryption;
+}
+
+/**
+ * @brief Find the first wifi-iface section in @a all_sections that matches
+ *        bSTA criteria: mode=sta, multi_ap=BSS_TYPE_BACKHAUL, and whose
+ *        associated wifi-device band matches @a radio_band.
+ *
+ * @return Section name of the first match, or empty string if none found.
+ */
+static std::string find_bsta_iface_section(const std::vector<std::string> &all_sections,
+                                           const std::string &radio_band)
+{
+    const std::string pkg  = "wireless";
+    const std::string type = "wifi-iface";
+    for (const auto &sec : all_sections) {
+        std::string mode_val;
+        if (!uci_get_option(pkg, type, sec, "mode", mode_val) || mode_val != "sta") {
+            continue;
+        }
+        std::string multi_ap_val;
+        uci_get_option(pkg, type, sec, "multi_ap", multi_ap_val);
+        if (multi_ap_val != std::to_string(beerocks::eBssType::BSS_TYPE_BACKHAUL)) {
+            continue;
+        }
+        std::string device_val;
+        if (!uci_get_option(pkg, type, sec, "device", device_val)) {
+            continue;
+        }
+        if (get_refined_band(device_val) == radio_band) {
+            return sec;
+        }
+    }
+    return "";
+}
+
+
 bool bpl_cfg_set_sta_vap_bh_credentials(const std::string &iface,
-                                        const son::wireless_utils::sBssInfoConf &configuration)
+                                        const son::wireless_utils::sBssInfoConf &configuration,
+                                        const std::string &radio_band)
 {
-    // Find all "wireless.wifi-iface" sections in UCI configuration
+    LOG(DEBUG) << "bpl_cfg_set_sta_vap_bh_credentials: iface=" << iface
+               << " radio_band=" << radio_band;
+
     const std::string package_name = "wireless";
     const std::string section_type = "wifi-iface";
     
-    // Get all wifi-iface sections
+
+    // Holds the wifi-mld section name associated with the matched wifi-iface section
+    std::string matched_mld_section_name;
+
+    // Get all wifi-iface sections once; reused for matching and disabled-option cloning
     std::vector<std::string> all_sections;
     if (!uci_get_all_sections(package_name, section_type, all_sections)) {
         LOG(ERROR) << "Failed to get all wifi-iface sections";
         return false;
     }
-    
-    // Lambda to get encryption string
-    auto get_encryption = [](WSC::eWscAuth authentication_type, WSC::eWscEncr encryption_type) {
-            std::string encryption = "none";
-            if (authentication_type == WSC::eWscAuth::WSC_AUTH_WPA2PSK) {
-                   encryption = "psk2";
-                   if (encryption_type == WSC::eWscEncr::WSC_ENCR_TKIP) {
-                           encryption += "+tkip";
-                   } else if (encryption_type == WSC::eWscEncr::WSC_ENCR_AES) {
-                           encryption += "+aes";
-                   }
-            } else if (authentication_type == WSC::eWscAuth::WSC_AUTH_SAE) {
-                   encryption = "sae";
-            } else if (authentication_type == (WSC::eWscAuth::WSC_AUTH_SAE|WSC::eWscAuth::WSC_AUTH_WPA2PSK)) {
-                   encryption = "sae-mixed";
-            }
-            return encryption;
-    };
-    
-    // Prepare options to write
+
+    // Prepare credential options using the shared get_encryption_string helper
     OptionsUnorderedMap options;
-    options["ssid"] = configuration.ssid;
-    options["encryption"] = get_encryption(configuration.authentication_type, configuration.encryption_type);
+    options["ssid"]       = configuration.ssid;
+    options["encryption"] = get_encryption_string(configuration.authentication_type,
+                                                   configuration.encryption_type);
 
-    // Set sae_pwe if encryption is SAE
     if (options["encryption"] == "sae") {
-           options["sae_pwe"] = "2";
-        options["sae"] = "1";
+        options["sae_pwe"] = "2";
+        options["sae"]     = "1";
+
     }
     if (options["encryption"] == "sae-mixed") {
-                options["sae_pwe"] = "2";
+        options["sae_pwe"] = "2";
+
     }
-    options["key"] = configuration.network_key;
-    
-    // Track if we updated at least one section
-    bool updated_any = false;
+    options["key"]      = configuration.network_key;
+    options["multi_ap"] = "1";
+    options["wds"]      = "1";
+
+    bool updated_any    = false;
+
     int sections_updated = 0;
-    
-    // Iterate through all sections and update those that match our criteria
+
+    // Update every wifi-iface section that matches: mode=sta, multi_ap=BSS_TYPE_BACKHAUL,
+    // and whose associated wifi-device band matches radio_band.
+
     for (const auto &section_name : all_sections) {
         std::string mode_value;
-        std::string ifname_value;
-        
-        // Get mode option
         if (!uci_get_option(package_name, section_type, section_name, "mode", mode_value)) {
             LOG(DEBUG) << "Failed to get 'mode' from section " << section_name;
             continue;
         }
+
+        std::string ifname_value;
         
-        // Get ifname option
         if (!uci_get_option(package_name, section_type, section_name, "ifname", ifname_value)) {
             LOG(DEBUG) << "Failed to get 'ifname' from section " << section_name;
             continue;
         }
-        
-        // Check if this section matches our criteria: mode=sta and ifname matches
-        if (mode_value == "sta" && ifname_value == iface) {
-            // Write UCI options in this matching section
-            if (!uci_set_section(package_name, section_type, section_name, options, true)) {
-                LOG(ERROR) << "Failed to set wireless configuration for interface " << iface 
-                          << " at section " << section_name;
-                continue;
-            }
-            
-            LOG(DEBUG) << "Successfully updated bh credentials for sta interface " << iface 
-                      << " (ifname=" << ifname_value << ") at section " << section_name;
-            updated_any = true;
-            sections_updated++;
+
+        std::string multi_ap_value;
+        uci_get_option(package_name, section_type, section_name, "multi_ap", multi_ap_value);
+
+        std::string device_value;
+        if (!uci_get_option(package_name, section_type, section_name, "device", device_value)) {
+            LOG(DEBUG) << "Failed to get 'device' from section " << section_name;
+            continue;
+         }
+
+        // Use get_refined_band helper (handles 5g → 5gl/5gh refinement)
+        std::string radio_band_value = get_refined_band(device_value);
+        if (radio_band_value.empty()) {
+            LOG(DEBUG) << "Failed to get 'band' from wifi-device section " << device_value;
+            continue;
+        }
+
+        if (mode_value != "sta" ||
+            multi_ap_value != std::to_string(beerocks::eBssType::BSS_TYPE_BACKHAUL) ||
+            radio_band_value != radio_band) {
+            continue;
         }
+
+        if (!uci_set_section(package_name, section_type, section_name, options, true)) {
+            LOG(ERROR) << "Failed to set wireless configuration for interface " << iface
+                       << " at section " << section_name;
+            continue;
+        }
+
+        LOG(DEBUG) << "Successfully updated bh credentials for sta interface " << iface
+                   << " (ifname=" << ifname_value << ", radio_band=" << radio_band_value
+                   << ") at section " << section_name;
+
+        // Track the associated wifi-mld section name (last match wins)
+        std::string mld_option_value;
+        if (uci_get_option(package_name, section_type, section_name, "mld", mld_option_value) &&
+            !mld_option_value.empty()) {
+            matched_mld_section_name = mld_option_value;
+            LOG(DEBUG) << "Found associated wifi-mld section: " << matched_mld_section_name
+                       << " from wifi-iface section " << section_name;
+        }
+
+        updated_any = true;
+        sections_updated++;
     }
     
     if (!updated_any) {
-        LOG(WARNING) << "No matching wifi-iface sections found for interface " << iface 
-                    << " with mode=sta";
-        return false;
+        LOG(INFO) << "No matching wifi-iface section found for radio_band=" << radio_band
+                  << ", creating new section for sta interface " << iface;
+
+        // Find the wifi-device section whose band matches radio_band using get_refined_band
+        std::string matching_device;
+        std::vector<std::string> all_device_sections;
+        if (uci_get_all_sections(package_name, "wifi-device", all_device_sections)) {
+            for (const auto &dev_section : all_device_sections) {
+                if (get_refined_band(dev_section) == radio_band) {
+                    matching_device = dev_section;
+                    break;
+                }
+            }
+        }
+
+        if (matching_device.empty()) {
+            LOG(ERROR) << "Could not find wifi-device section for radio_band=" << radio_band
+                       << ", cannot create new section for interface " << iface;
+            return false;
+        }
+
+        std::string new_section;
+        int iface_index = 0;
+        do {
+            new_section = "iface_wlan" + std::to_string(iface_index++);
+        } while (uci_section_exists(package_name, section_type, new_section));
+
+        if (!uci_add_section(package_name, section_type, new_section, true)) {
+            LOG(ERROR) << "Failed to add new wifi-iface section for interface " << iface;
+            return false;
+        }
+
+        options["mode"]   = "sta";
+        options["device"] = matching_device;
+
+        // Find a reference section by identifying which section has mode="sta".
+        // The device option is NOT used as the selection criterion here.
+        std::string ref_sec;
+        for (const auto &s : all_sections) {
+            std::string mode_val;
+            if (uci_get_option(package_name, section_type, s, "mode", mode_val) &&
+                mode_val == "sta") {
+                ref_sec = s;
+                break;
+            }
+        }
+
+        // Clone ifname from the reference section if available.
+        // If ref_sec is empty, ifname is left unset (empty).
+        if (!ref_sec.empty()) {
+            std::string ref_ifname;
+            if (uci_get_option(package_name, section_type, ref_sec, "ifname", ref_ifname) &&
+                !ref_ifname.empty()) {
+                options["ifname"] = ref_ifname;
+            }
+        }
+
+        // Clone "disabled" from the reference section, or set to 1 if no reference exists.
+        if (!ref_sec.empty()) {
+            std::string ref_disabled;
+            if (uci_get_option(package_name, section_type, ref_sec, "disabled", ref_disabled)) {
+                options["disabled"] = ref_disabled;
+            }
+        } else {
+            options["disabled"] = "1";
+        }
+
+        if (!uci_set_section(package_name, section_type, new_section, options, true)) {
+            LOG(ERROR) << "Failed to set wireless configuration for new section " << new_section
+                       << " (interface " << iface << ")";
+            return false;
+        }
+
+        LOG(INFO) << "Created new wifi-iface section " << new_section
+                  << " for sta interface " << iface << " (radio_band=" << radio_band
+                  << ", device=" << matching_device << ")";
+        return true;
     }
-    
-    LOG(INFO) << "Successfully updated " << sections_updated 
+
+    LOG(INFO) << "Successfully updated " << sections_updated
               << " wifi-iface section(s) with bh credentials for sta interface " << iface;
-    
-    // Now update wifi-mld sections that match the ifname
+
+    // Update the wifi-mld section identified from the matched wifi-iface section's mld option
     const std::string mld_section_type = "wifi-mld";
-    std::vector<std::string> all_mld_sections;
-    if (!uci_get_all_sections(package_name, mld_section_type, all_mld_sections)) {
-        LOG(DEBUG) << "Failed to get all wifi-mld sections or no wifi-mld sections exist";
-        // Not a critical error, wifi-mld sections may not exist
-        return true;
-    }
-    
-    // Prepare options for wifi-mld (no encryption field in mld section)
     OptionsUnorderedMap mld_options;
-    mld_options["ssid"] = configuration.ssid;
-    mld_options["key"] = configuration.network_key;
-    
-    int mld_sections_updated = 0;
-    
-    // Iterate through all mld sections and update those that match the ifname
-    for (const auto &mld_section_name : all_mld_sections) {
-        std::string mld_ifname_value;
-        
-        // Get ifname option
-        if (!uci_get_option(package_name, mld_section_type, mld_section_name, "ifname", mld_ifname_value)) {
-            LOG(DEBUG) << "Failed to get 'ifname' from mld section " << mld_section_name;
-            continue;
+    mld_options["ssid"]     = configuration.ssid;
+    mld_options["key"]      = configuration.network_key;
+    mld_options["multi_ap"] = "1";
+    mld_options["wds"]      = "1";
+
+    if (!matched_mld_section_name.empty()) {
+        if (uci_section_exists(package_name, mld_section_type, matched_mld_section_name)) {
+            if (!uci_set_section(package_name, mld_section_type, matched_mld_section_name,
+                                 mld_options, true)) {
+                LOG(ERROR) << "Failed to set wireless configuration for interface " << iface
+                           << " at mld section " << matched_mld_section_name;
+            } else {
+                LOG(INFO) << "Successfully updated bh credentials for mld section "
+                          << matched_mld_section_name;
+            }
+        } else {
+            LOG(DEBUG) << "wifi-mld section " << matched_mld_section_name
+                       << " does not exist, skipping";
+        }
+    } else {
+        LOG(DEBUG) << "No associated wifi-mld section found for interface " << iface;
+    }
+
+    return true;
+}
+
+
+eBstaSupplicantRestartStatus bpl_cfg_update_bsta_mld_grp(const std::string &radio_band,
+                                                         int8_t mld_unit)
+{
+    LOG(DEBUG) << "bpl_cfg_update_bsta_mld_grp: radio_band=" << radio_band
+               << " mld_unit=" << int(mld_unit);
+
+    const std::string package_name  = "wireless";
+    const std::string section_type  = "wifi-iface";
+    const std::string mld_sect_type = "wifi-mld";
+
+    // Get all wifi-iface sections once; shared by both the mld_unit=-1 path and the
+    // normal path via find_bsta_iface_section.
+    std::vector<std::string> all_sections;
+    if (!uci_get_all_sections(package_name, section_type, all_sections)) {
+        LOG(ERROR) << "bpl_cfg_update_bsta_mld_grp: failed to get all wifi-iface sections";
+        return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
+    }
+
+    // Use find_bsta_iface_section to locate the matching bSTA wifi-iface section
+    // (mode=sta, multi_ap=BSS_TYPE_BACKHAUL, band matches radio_band).
+    const std::string matched_section = find_bsta_iface_section(all_sections, radio_band);
+
+    // If mld_unit == -1, remove the bSTA section (or just its mld option when it is the
+    // only section referencing mld0).
+    if (mld_unit == -1) {
+        if (matched_section.empty()) {
+            LOG(DEBUG) << "bpl_cfg_update_bsta_mld_grp: mld_unit=-1, no matching wifi-iface "
+                       << "section found for radio_band=" << radio_band << ", nothing to delete";
+            return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
+        }
+
+        // If this section has mld="mld0", check whether it is the only such section.
+        // If it is the only one, delete the wifi-mld section and remove the mld option
+        // from the wifi-iface section instead of deleting the section itself.
+        std::string current_mld_value;
+        uci_get_option(package_name, section_type, matched_section, "mld", current_mld_value);
+
+        if (current_mld_value == "mld0") {
+            // Count other wifi-iface sections with mld="mld0" AND mode="sta"
+            int mld0_sta_count = 0;
+            for (const auto &other_sec : all_sections) {
+                if (other_sec == matched_section) {
+                    continue;
+                }
+                std::string other_mode, other_mld;
+                if (uci_get_option(package_name, section_type, other_sec, "mode", other_mode) &&
+                    other_mode == "sta" &&
+                    uci_get_option(package_name, section_type, other_sec, "mld", other_mld) &&
+                    other_mld == "mld0") {
+                    mld0_sta_count++;
+                }
+            }
+
+            if (mld0_sta_count == 0) {
+                LOG(INFO) << "bpl_cfg_update_bsta_mld_grp: mld_unit=-1, section "
+                          << matched_section
+                          << " is the only wifi-iface with mld=mld0 and mode=sta; "
+                          << "removing wifi-mld section mld0 and mld option from wifi-iface";
+
+                if (uci_section_exists(package_name, mld_sect_type, "mld0")) {
+                    if (!uci_delete_section(package_name, mld_sect_type, "mld0", true)) {
+                        LOG(ERROR) << "bpl_cfg_update_bsta_mld_grp: failed to delete "
+                                   << "wifi-mld section mld0";
+                        return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
+                    }
+                    LOG(INFO) << "bpl_cfg_update_bsta_mld_grp: deleted wifi-mld section mld0";
+                }
+
+                // Remove the "mld" option by setting it to "" (uci_set_section removes it)
+                OptionsUnorderedMap remove_mld_opts;
+                remove_mld_opts["mld"] = "";
+                if (!uci_set_section(package_name, section_type, matched_section,
+                                     remove_mld_opts, true)) {
+                    LOG(ERROR) << "bpl_cfg_update_bsta_mld_grp: failed to remove mld option "
+                               << "from section " << matched_section;
+                    return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
+                }
+                LOG(INFO) << "bpl_cfg_update_bsta_mld_grp: removed mld option from "
+                          << "wifi-iface section " << matched_section;
+                return eBstaSupplicantRestartStatus::SUPP_RST_REQUIRED;
+            }
+        }
+
+        // Delete the matching section (either it has no mld="mld0", or other sections
+        // with mld="mld0" and mode="sta" will remain after deletion).
+        LOG(INFO) << "bpl_cfg_update_bsta_mld_grp: mld_unit=-1, deleting wifi-iface section "
+                  << matched_section << " for radio_band=" << radio_band;
+        if (!uci_delete_section(package_name, section_type, matched_section, true)) {
+            LOG(ERROR) << "bpl_cfg_update_bsta_mld_grp: failed to delete section "
+                       << matched_section;
+            return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
         }
         
-        // Check if this mld section matches our ifname
-        if (mld_ifname_value == iface) {
-            // Write UCI options in this matching mld section
-            if (!uci_set_section(package_name, mld_section_type, mld_section_name, mld_options, true)) {
-                LOG(ERROR) << "Failed to set wireless configuration for interface " << iface 
-                          << " at mld section " << mld_section_name;
-                continue;
+        return eBstaSupplicantRestartStatus::SUPP_RST_REQUIRED;
+    }
+
+    // bSTA MLD unit must be 0 for normal operation
+    if (mld_unit != 0) {
+        LOG(WARNING) << "bpl_cfg_update_bsta_mld_grp: mld_unit=" << int(mld_unit)
+                     << " is not 0, using 0";
+        mld_unit = 0;
+    }
+
+    // Derive the mld section name from mld_unit: "mld0"
+    const std::string mld_name = "mld" + std::to_string(mld_unit);
+
+    // If no valid wifi-iface section was found, nothing to do
+    if (matched_section.empty()) {
+        LOG(DEBUG) << "bpl_cfg_update_bsta_mld_grp: no matching wifi-iface section found for "
+                   << "radio_band=" << radio_band;
+        return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
+    }
+
+    // Ensure the mld option in the matched wifi-iface section is set to mld_name
+    std::string current_mld_value;
+    bool has_mld = uci_get_option(package_name, section_type, matched_section, "mld",
+                                  current_mld_value);
+    if (!has_mld || current_mld_value.empty()) {
+        LOG(INFO) << "bpl_cfg_update_bsta_mld_grp: section " << matched_section
+                  << " has no mld option, setting to " << mld_name;
+        OptionsUnorderedMap mld_set_options;
+        mld_set_options["mld"] = mld_name;
+        if (!uci_set_section(package_name, section_type, matched_section, mld_set_options,
+                             true)) {
+            LOG(ERROR) << "bpl_cfg_update_bsta_mld_grp: failed to set mld option in section "
+                       << matched_section;
+            return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
+        }
+    } else if (current_mld_value != mld_name) {
+        LOG(INFO) << "bpl_cfg_update_bsta_mld_grp: updating section " << matched_section
+                  << " mld option from '" << current_mld_value << "' to '" << mld_name << "'";
+        OptionsUnorderedMap upd_options;
+        upd_options["mld"] = mld_name;
+        if (!uci_set_section(package_name, section_type, matched_section, upd_options, true)) {
+            LOG(ERROR) << "bpl_cfg_update_bsta_mld_grp: failed to update mld option in section "
+                       << matched_section;
+            return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
+        }
+    }
+
+    // --- wifi-mld section check ---
+    // Keys to sync from wifi-iface to wifi-mld.
+    // Keys not present in the wifi-iface section are silently skipped.
+    static const char * const mld_sync_keys[] = {
+        "ssid", "encryption", "key", "sae", "sae_pwe",
+        "ifname", "multi_ap", "wps_pushbutton", nullptr
+    };
+
+    if (uci_section_exists(package_name, mld_sect_type, mld_name)) {
+        // wifi-mld section exists — compare each sync key and update if changed.
+        OptionsUnorderedMap updated_opts;
+        bool any_changed = false;
+
+        for (int i = 0; mld_sync_keys[i] != nullptr; ++i) {
+            const std::string key = mld_sync_keys[i];
+            std::string iface_val;
+            if (!uci_get_option(package_name, section_type, matched_section, key, iface_val)) {
+               continue;
+           }
+            std::string mld_val;
+            uci_get_option(package_name, mld_sect_type, mld_name, key, mld_val);
+            if (iface_val != mld_val) {
+                LOG(DEBUG) << "bpl_cfg_update_bsta_mld_grp: wifi-mld option '" << key
+                           << "' differs (mld='" << mld_val << "' iface='" << iface_val << "')";
+                updated_opts[key] = iface_val;
+                any_changed       = true;
             }
+        }
+
+        if (any_changed) {
+            LOG(INFO) << "bpl_cfg_update_bsta_mld_grp: updating wifi-mld section " << mld_name
+                      << " with changed options";
+            if (!uci_set_section(package_name, mld_sect_type, mld_name, updated_opts, true)) {
+                LOG(ERROR) << "bpl_cfg_update_bsta_mld_grp: failed to update wifi-mld section "
+                           << mld_name;
+            }
+            return eBstaSupplicantRestartStatus::SUPP_RST_REQUIRED;
             
-            LOG(DEBUG) << "Successfully updated bh credentials for mld section " << mld_section_name
-                      << " (ifname=" << mld_ifname_value << ")";
-            mld_sections_updated++;
         }
+
+        LOG(DEBUG) << "bpl_cfg_update_bsta_mld_grp: wifi-mld section " << mld_name
+                   << " is in sync with wifi-iface section";
+        return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
     }
-    
-    if (mld_sections_updated > 0) {
-        LOG(INFO) << "Successfully updated " << mld_sections_updated 
-                  << " wifi-mld section(s) with bh credentials for interface " << iface;
+
+    // wifi-mld section does not exist — create it from the matched wifi-iface section
+    LOG(INFO) << "bpl_cfg_update_bsta_mld_grp: creating wifi-mld section " << mld_name
+              << " from wifi-iface section " << matched_section;
+
+    if (!uci_add_section(package_name, mld_sect_type, mld_name, true)) {
+        LOG(ERROR) << "bpl_cfg_update_bsta_mld_grp: failed to create wifi-mld section "
+                   << mld_name;
+        return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
+    }
+
+    OptionsUnorderedMap new_mld_opts;
+    for (int i = 0; mld_sync_keys[i] != nullptr; ++i) {
+        std::string val;
+        if (uci_get_option(package_name, section_type, matched_section, mld_sync_keys[i], val)) {
+            new_mld_opts[mld_sync_keys[i]] = val;
+        }
+    }
+
+    // Clone "disabled" from the reference wifi-iface section, or set to 1 if none exists.
+    if (!matched_section.empty()) {
+        std::string ref_disabled;
+        if (uci_get_option(package_name, section_type, matched_section, "disabled", ref_disabled)) {
+            new_mld_opts["disabled"] = ref_disabled;
+        }
     } else {
-        LOG(DEBUG) << "No matching wifi-mld sections found for interface " << iface;
+        new_mld_opts["disabled"] = "1";
+    }
+    if (!uci_set_section(package_name, mld_sect_type, mld_name, new_mld_opts, true)) {
+        LOG(ERROR) << "bpl_cfg_update_bsta_mld_grp: failed to populate wifi-mld section "
+                   << mld_name;
+        return eBstaSupplicantRestartStatus::SUPP_RST_NOT_REQUIRED;
     }
-    
-    return true;
-}
 
+    LOG(INFO) << "bpl_cfg_update_bsta_mld_grp: created wifi-mld section " << mld_name;
+    return eBstaSupplicantRestartStatus::SUPP_RST_REQUIRED;
+}
 
 bool bpl_cfg_set_wifi_iface_teardown(const std::string &iface, const std::string &radio_id, bool is_mlo)
 {
diff --git a/framework/platform/bpl/include/bpl/bpl_cfg.h b/framework/platform/bpl/include/bpl/bpl_cfg.h
index d42fcf4fc..265114b16 100644
--- a/framework/platform/bpl/include/bpl/bpl_cfg.h
+++ b/framework/platform/bpl/include/bpl/bpl_cfg.h
@@ -1545,8 +1545,35 @@ bool cfg_update_prplmesh_option(const std::string &package_name,
                             const std::string &new_value);
 
 bool bpl_cfg_set_sta_vap_bh_credentials(const std::string &iface,
-                                  const son::wireless_utils::sBssInfoConf &configuration);
+                                  const son::wireless_utils::sBssInfoConf &configuration,
+                                  const std::string &radio_band);
 
+/**
+ * @brief Return codes for bpl_cfg_update_bsta_mld_grp indicating whether
+ * the wpa_supplicant needs to be restarted to pick up the new MLD group.
+ */
+enum class eBstaSupplicantRestartStatus {
+    SUPP_RST_NOT_REQUIRED = 0, /**< mld option already matches, no restart needed */
+    SUPP_RST_REQUIRED     = 1, /**< mld option was updated, restart required */
+};
+
+/**
+ * @brief Update the MLD group option in the UCI wifi-iface section for the bSTA.
+ *
+ * Finds the wifi-iface section whose band matches @p radio_band and has
+ * mode=sta and multi_ap=1, then checks/updates its "mld" option to the
+ * value derived from @p mld_unit (e.g. "mld0").
+ *
+ * @param [in] radio_band  Radio band string (e.g. "2g", "5gl", "5gh", "6g").
+ * @param [in] mld_unit    MLD unit number (must be 0 for bSTA).
+ * @return eBstaSupplicantRestartStatus::supplicant_restart_required if the mld
+ *         option was updated (restart needed), or
+ *         eBstaSupplicantRestartStatus::supplicant_restart_not_required if it
+ *         was already correct.
+ */
+eBstaSupplicantRestartStatus bpl_cfg_update_bsta_mld_grp(const std::string &radio_band,
+                                                                int8_t mld_unit);
+ 
 /**
  * @brief Gets the MLD ID for a given interface from the wifi-mld UCI section.
  *
-- 
2.34.1

